File: app/src/main/java/com/example/alarm/SignalActivity.java
package com.example.alarm;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.KeyEvent;
import android.view.WindowManager;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.IntentCompat;
import com.example.alarm.databinding.ActivitySignalBinding;
import com.example.alarm.model.Settings;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class SignalActivity extends AppCompatActivity {

    private ActivitySignalBinding binding;
    private boolean isHomePressed = false;
    private Handler homePressResetHandler;
    private Runnable homePressResetRunnable;

    public static String APP_PREFERENCES = "APP_PREFERENCES";
    public static String PREF_THEME = "PREF_THEME";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        int themeNumber = getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE).getInt(PREF_THEME, 0);
        switch (themeNumber) {
            case 0: setTheme(R.style.Theme_Alarm); break;
            case 1: setTheme(R.style.Theme1); break;
            case 2: setTheme(R.style.Theme2); break;
            case 3: setTheme(R.style.Theme3); break;
            case 4: setTheme(R.style.Theme4); break;
            case 5: setTheme(R.style.Theme5); break;
            case 6: setTheme(R.style.Theme6); break;
            case 7: setTheme(R.style.Theme7); break;
            case 8: setTheme(R.style.Theme8); break;
            default: setTheme(R.style.Theme_Alarm); break;
        }
        super.onCreate(savedInstanceState);
        binding = ActivitySignalBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
                getWindow().addFlags(
                WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON |
                        WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD |
                        WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED |
                        WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
        );

        String alarmName = getIntent().getStringExtra("alarmName");
        long alarmId = getIntent().getLongExtra("alarmId", 0);
        Settings settings = IntentCompat.getParcelableExtra(getIntent(), "settings", Settings.class);

        homePressResetHandler = new Handler(Looper.getMainLooper());

        homePressResetRunnable = new Runnable() {
            @Override
            public void run() {
                isHomePressed = false;
            }
        };
        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.fragmentContainer2, new SignalFragment(alarmName, alarmId, settings))
                    .commit();
        }
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (getSupportFragmentManager().findFragmentById(R.id.fragmentContainer2) instanceof SignalFragment) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                ((SignalFragment) getSupportFragmentManager().findFragmentById(R.id.fragmentContainer2)).dropAndRepeatFragment();
            }
        }
        return true;
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (getSupportFragmentManager().findFragmentById(R.id.fragmentContainer2) instanceof SignalFragment) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                ((SignalFragment) getSupportFragmentManager().findFragmentById(R.id.fragmentContainer2)).dropAndRepeatFragment();
            }
        }
        return true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        isHomePressed = true;
        homePressResetHandler.removeCallbacks(homePressResetRunnable);
        homePressResetHandler.postDelayed(homePressResetRunnable, 1000);
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (isHomePressed) {
            SignalFragment fragment = (SignalFragment) getSupportFragmentManager().findFragmentById(R.id.fragmentContainer2);
            if (fragment != null) {
                fragment.dropAndRepeatFragment();
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        homePressResetHandler.removeCallbacks(homePressResetRunnable);
    }
}

File: app/src/main/java/com/example/alarm/SignalActivity.kt
package com.example.alarm

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.IntentCompat
import androidx.lifecycle.lifecycleScope
import com.example.alarm.databinding.ActivitySignalBinding
import com.example.alarm.model.Settings
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SignalActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySignalBinding
    private var isHomePressed = false
    private var homePressResetJob: Job? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        val preferences = getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE)
        val themeNumber = preferences.getInt(PREF_THEME, 0)
        when(themeNumber) {
            0 -> setTheme(R.style.Theme_Alarm)
            1 -> setTheme(R.style.Theme1)
            2 -> setTheme(R.style.Theme2)
            3 -> setTheme(R.style.Theme3)
            4 -> setTheme(R.style.Theme4)
            5 -> setTheme(R.style.Theme5)
            6 -> setTheme(R.style.Theme6)
            7 -> setTheme(R.style.Theme7)
            8 -> setTheme(R.style.Theme8)
            else -> setTheme(R.style.Theme_Alarm)
        }
        super.onCreate(savedInstanceState)
        binding = ActivitySignalBinding.inflate(layoutInflater).also { setContentView(it.root) }
        setContentView(binding.root)
        @Suppress("DEPRECATION")
        window.addFlags(
            WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON or
                    WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD or
                    WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED or
                    WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
        )
        val alarmName = intent.getStringExtra("alarmName") ?: ""
        val alarmId = intent.getLongExtra("alarmId", 0)
        val settings = IntentCompat.getParcelableExtra(intent, "settings", Settings::class.java)
        Log.d("testSignActivity", settings.toString())
        if (savedInstanceState == null) {
            supportFragmentManager
                .beginTransaction()
                .replace(R.id.fragmentContainer2, SignalFragment(alarmName, alarmId, settings))
                .commit()
        }
    }
    override fun onKeyUp(keyCode: Int, event: KeyEvent?): Boolean {
        if (supportFragmentManager.findFragmentById(R.id.fragmentContainer2) is SignalFragment) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                (supportFragmentManager.findFragmentById(R.id.fragmentContainer2) as SignalFragment).dropAndRepeatFragment()
            }
        }
        return true
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (supportFragmentManager.findFragmentById(R.id.fragmentContainer2) is SignalFragment) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                (supportFragmentManager.findFragmentById(R.id.fragmentContainer2) as SignalFragment).dropAndRepeatFragment()
            }
        }
        return true
    }

    override fun onPause() {
        super.onPause()
        isHomePressed = true
        homePressResetJob?.cancel()
        homePressResetJob = lifecycleScope.launch {
            delay(1000)
            isHomePressed = false
        }
    }

    override fun onStop() {
        super.onStop()
        if (isHomePressed) {
            (supportFragmentManager.findFragmentById(R.id.fragmentContainer2) as SignalFragment).dropAndRepeatFragment()
        }
    }
}